File: ios/Classes/Bridges/RestorationResultBridge.swift
import Flutter
import SuperwallKit

public class RestorationResultBridge: BridgeInstance {
  var restorationResult: RestorationResult {
    assertionFailure("Subclasses must implement")
    return .failed(RestorationResultError(message: "Subclasses must implement restorationResult"))
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterMethodNotImplemented)
  }
}

public class RestorationResultRestoredBridge: RestorationResultBridge {
  override class func bridgeClass() -> BridgeClass { return "RestorationResultRestoredBridge" }
  override var restorationResult: RestorationResult { return .restored }
}

public class RestorationResultFailedBridge: RestorationResultBridge {
  override class func bridgeClass() -> BridgeClass { return "RestorationResultFailedBridge" }
  override var restorationResult: RestorationResult { .failed(error) }

  let error: Error

  required init(bridgeId: BridgeId, initializationArgs: [String : Any]? = nil) {
    guard let error = initializationArgs?["error"] as? String else {
      fatalError("Attempting to create `RestorationResultFailedBridge` without providing `error`.")
    }

    self.error = RestorationResultError(message: error)
    super.init(bridgeId: bridgeId, initializationArgs: initializationArgs)
  }
}

struct RestorationResultError: Error {
  let message: String
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/bridges/RestorationResultBridge.kt
package com.superwall.superwallkit_flutter.bridges

import android.content.Context
import com.superwall.sdk.delegate.RestorationResult
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

abstract class RestorationResultBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : BridgeInstance(context, bridgeId, initializationArgs), MethodChannel.MethodCallHandler {

    abstract val restorationResult: RestorationResult

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        result.notImplemented()
    }
}

class RestorationResultRestoredBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : RestorationResultBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "RestorationResultRestoredBridge" }
    override val restorationResult: RestorationResult = RestorationResult.Restored()
}

class RestorationResultFailedBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : RestorationResultBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "RestorationResultFailedBridge" }
    override val restorationResult: RestorationResult

    init {
        val errorMessage = initializationArgs?.get("error") as? String
            ?: throw IllegalArgumentException("Attempting to create `RestorationResultFailedBridge` without providing `error`.")
        restorationResult = RestorationResult.Failed(RestorationResultError(errorMessage))
    }
}

data class RestorationResultError(override val message: String) : Throwable()
